Python/chapter02/2.4 - Partition/miguel_2.4_soln.py

```diff
@@ -0,0 +1,245 @@
+    def __init__(self, d: int):
+        self.data = d
+        self.next = None
+
+    def __repr__(self):
+        return self.__str__()
+
+    def __str__(self):
+        return '<Node Value: {}>'.format(self.data)
+
+    def __eq__(self, other: object):
+        if not isinstance(other, Node):
+            return NotImplemented
+        return self.data == other.data
+
+
+class LinkedList:
+    def __init__(self, *numbers: int):
+        self.head = None
+        self.tail = None
+        self.size = 0
+        for num in numbers:
+            self.append_to_tail(num)
+
+    def append_to_tail(self, d: int) -> None:
+        if self.head is None:
+            self.head = Node(d)
+            self.tail = self.head
+        else:
+            end = Node(d)
+            self.tail.next = end
+            self.tail = end
+        self.size += 1
+
+    def append_to_head(self, d: int) -> None:
+        new_head = Node(d)
+        new_head.next = self.head
+        self.head = new_head
+        self.size += 1
+
+    def get_node_at(self, index: int) -> Node:
+        if index < 0 or index >= self.size:
+            raise IndexError('list index out of range')
+        n = self.head
+        for i in range(self.size):
+            if i == index:
+                return n
+            n = n.next
+
+    def __repr__(self):
+        return self.__str__()
+
+    def __str__(self):
+        if self.head is None:
+            return '<empty>'
+        ll = []
+        n = self.head
+        while n.next is not None:
+            ll.append('{} -> '.format(n.data))
+            n = n.next
+        ll.append(str(n.data))
+        return ''.join(ll)
+
+    def __eq__(self, other: object):
+        if not isinstance(other, LinkedList):
+            return NotImplemented
+        a = self.head
+        b = other.head
+        while a is not None and b is not None:
+            if a.data != b.data:
+                return False
+            # otherwise, advance both pointers
+            a = a.next
+            b = b.next
+        return a is None and b is None
+
+
+def check_partitioned(ll: LinkedList, pivot: int) -> bool:
+    """
+    Will check if a linked list is partitioned
+    around a value such that all values less
+    then the partition value come before all values
+    greater than or equal to the partition value
+    in the linked list.
+    Order doesn't matter, as long as the values
+    < partition come before values >= partition.
+    Runtime: O(n)
+    Space Complexity: O(1)
+    :param ll: an input linked list
+    :param pivot: a number to partition around
+    :return:
+    """
+    # counter for keeping track of how many times
+    # we change from < pivot to >= pivot
+    pivot_change_count = 0
+    n = ll.head
+    while n is not None:
+        if n.data >= pivot:
+            pivot_change_count += 1
+            break
+        n = n.next
+    while n is not None:
+        if n.data < pivot:
+            pivot_change_count += 1
```

Comment:
`pivot_change_count` cannot be less than 1 here, so whenever this line is reached, you can simply `return False`. that means that L126 can never be more than 1, so you can change that one to `return True`. finally, you are not reading `pivot_change_count` anywhere, so you can remove any mentions of that variable.

which ultimately leads to code that is shorter, easier to understand _and_ has less repetition!